### PR TITLE
feat: add the validate utils

### PR DESCRIPTION
**Pull request by duguorong009**
## What kind of change does this PR introduce?
Add the `validate` utils for getting the actual validation errors thrown while validating the json schema and instance.

## What is the current behavior?
Solves #14 

## What is the new behavior?
Receives the actual errors thrown, as `Vec<String>`

### CODE REVIEW

**Comment by olirice** [comment]
hi @duguorong009 please take a look at the CI errors when you get a minute
thanks!

**Comment by olirice** [comment]
sorry for the delay getting back to you, its been a hectic week
will leave a review nlt tomorrow

File: src/lib.rs [hunk 1]
```diff
@@ -29,6 +29,64 @@ fn jsonschema_is_valid(schema: Json) -> bool {
     }
 }
 
+#[pg_extern(immutable, strict)]
+fn validate_json_schema(schema: Json, instance: Json) -> bool {
```
Comment (olirice) [inline src/lib.rs:33]: From the linked issue I thought the goal was to produce functions that return an array of strings that described and errors that are present.

These two functions have identical signatures and behavior to `json_matches_schema` and `jsonb_matches_schema` but have raise notices that list the errors.

If that is the goal, why not update the existing two functions? e.g. is it a performance problem? if so, what is the penalty?

If it is something like performance, an optional argument to the existing functions like `on_error_raise` that takes one of `nothing` (default), `notice`, or `exception` would be preferred to adding new functions

Comment (duguorong009) [reply src/lib.rs]: Yes, there is performance problem.
From the doc, the `jsonschema::is_valid` is much faster than `validate` function.

Anyhow, I removed the new functions, and updated the existing functions as recommended.
Please re-check it. @olirice

**Comment by olirice** [comment]
@duguorong009 sorry for the back and forth

It's not clear to me what we're solving here. #14 requested exposing `validate()` which I understood to mean adding a function that would return an array of error strings.

Given that there is a performance hit for using  `validate` vs `is_valid` we're trading performance for hints.

But check constraints won't bubble up those hints, logs would need to be set to `verbose` to include hints, and supabase studio doesn't display hints (AO 2023-07). So the only time the hints would be visible is if a user interactively connected to the database. Due to that, I'm not sure this is an appropriate trade to make.

---

If you'd be interested in changing this to a single immutable function (json only, not jsonb)
```rust
fn jsonschema_validation_errors(schema: Json, instance: Json) -> Vec<String>
```
that returns an array of errors or an empty array (if there are no errors) I'd be happy to approve

thanks!

**Comment by duguorong009** [comment]
> @duguorong009 sorry for the back and forth
> 
> It's not clear to me what we're solving here. #14 requested exposing `validate()` which I understood to mean adding a function that would return an array of error strings.
> 
> Given that there is a performance hit for using `validate` vs `is_valid` we're trading performance for hints.
> 
> But check constraints won't bubble up those hints, logs would need to be set to `verbose` to include hints, and supabase studio doesn't display hints (AO 2023-07). So the only time the hints would be visible is if a user interactively connected to the database. Due to that, I'm not sure this is an appropriate trade to make.
> 
> If you'd be interested in changing this to a single immutable function (json only, not jsonb)
> 
> ```rust
> fn jsonschema_validation_errors(schema: Json, instance: Json) -> Vec<String>
> ```
> 
> that returns an array of errors or an empty array (if there are no errors) I'd be happy to approve
> 
> thanks!

Thanks for the detailed tips! @olirice

**Comment by olirice** [comment]
looks good

please include tests for 0 errors, 1 error, and multiple errors:

e.g.
```rs
// No errors
crate::jsonschema_validation_errors(
    Json(json!({ "maxLength": 4 })),
    Json(json!("foo")),
)

// One error
crate::jsonschema_validation_errors(
    Json(json!({ "maxLength": 4 })),
    Json(json!("123456789")),
)

// Multiple errors
crate::jsonschema_validation_errors(
    Json(json!(
    {
        "type": "object",
        "properties": {
            "foo": {
                "type": "string"
            },
            "bar": {
                "type": "number"
            },
            "baz": {
                "type": "boolean"
            },
            "additionalProperties": false,
            "required": ["foo", "bar", "baz"]
        }
    })),
    Json(json!({"foo": 1, "bar": [], "bat": true})),
)
```

File: src/lib.rs [hunk 2]
```diff
@@ -132,6 +145,54 @@ mod tests {
...
+                    "baz": {
+                        "type": "boolean"
+                    },
+                    "additionalProperties": false,
+                    "required": ["foo", "bar", "baz"]
+                }
+            })),
+            Json(json!({"foo": 1, "bar": [], "bat": true})),
+        );
+        assert!(errors.len() == 1);
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply src/lib.rs]: Thanks for comment!
Will update! @olirice

**Review by olirice — APPROVED** [review]

**Comment by olirice** [comment]
thanks!

**Pull request merged**